scripts/lldb/firestore.py

```diff
@@ -0,0 +1,250 @@
+        break
+
+      child = value.GetChildAtIndex(0)
+      if not child.IsValid():
+        break
+
+      if 'optional_internal' not in child.GetType().GetName():
+        break
+
+      value = child
+
+    # value should now point to the innermost absl::optional container type.
+    self.engaged = value.GetChildMemberWithName('engaged_')
+
+    if self.has_children():
+      self.data = value.GetChildMemberWithName('data_')
+
+    else:
+      self.data = None
+
+  def has_children(self):
+    return self.engaged.GetValueAsUnsigned(0) != 0
+
+  def num_children(self):
+    return 2 if self.has_children() else 1
+
+  def get_child_index(self, name):
+    if name == 'engaged_':
+      return 0
+    if name == 'data_':
+      return 1
+    return -1
+
+  def get_child_at_index(self, index):
+    if index == 0:
+      return self.engaged
+    if index == 1:
+      return self.data
+
+
+def AbseilOptional_SummaryProvider(value, params):
+  # Operates on the synthetic children above, calling has_children.
+  return 'engaged={0}'.format(format_bool(value.MightHaveChildren()))
+
+
+# model
+
+class DatabaseId_SynthProvider(ForwardingSynthProvider):
+  """Makes DatabaseId behave as if `*rep_` were inline."""
+  def delegate(self):
+    return deref_shared(self.value.GetChildMemberWithName('rep_'))
+
+
+def DatabaseId_SummaryProvider(value, params):
+  # Operates on the result of the SynthProvider; value is *rep_.
+  parts = [
+      get_string(value.GetChildMemberWithName('project_id')),
+      get_string(value.GetChildMemberWithName('database_id'))
+  ]
+  return format_string('/'.join(parts))
+
+
+def DocumentKey_SummaryProvider(value, params):
+  """Summarizes DocumentKey as if path_->segments_ were inline and a single
+  string."""
+  return deref_shared(value.GetChildMemberWithName('path_')).GetSummary()
+
+
+def ResourcePath_SummaryProvider(value, params):
+  """Summarizes ResourcePath as if segments_ is a single string."""
+
+  segments = value.GetChildMemberWithName('segments_')
+  count = segments.GetNumChildren()
+
+  segment_text = []
+  for i in range(0, count):
+    child = segments.GetChildAtIndex(i)
+    segment_text.append(get_string(child))
+
+  text = format_string('/'.join(segment_text))
+  return text
+
+
+# api
+
+def DocumentReference_SummaryProvider(value, params):
+  return value.GetChildMemberWithName('key_').GetSummary()
+
+
+def DocumentSnapshot_SummaryProvider(value, params):
+  return value.GetChildMemberWithName('internal_key_').GetSummary()
+
+
+# Objective-C
+
+def FIRDocumentReference_SummaryProvider(value, params):
+  return value.GetChildMemberWithName('_documentReference').GetSummary()
+
+
+def FIRDocumentSnapshot_SummaryProvider(value, params):
+  return value.GetChildMemberWithName('_snapshot').GetSummary()
+
+
+def get_string(value):
+  """Returns a Python string from the underlying LLDB SBValue."""
+  # TODO(wilhuff): This is gross hack. Actually use the SBData API to get this.
```

Comment:
Question: what's the reason to use a hack? If you expect it to exist for some time, it might be worthwhile to expand the comment a little bit to explain why the hack is in place.

Reply:
Pulling out the string contents requires dealing with the memory access API which is more than I wanted to bite off here. Instead, this gets lldb to format that into a Python string containing the summary and then evaluates the string as if it were a string literal to remote quotes and escape sequences. 

I've reformulated this note.